### PR TITLE
Add general transformer

### PR DESCRIPTION
**Pull request by pz-max**
## Changes proposed in this Pull Request
This is a staged PR and builds up on https://github.com/pypsa-meets-africa/pypsa-africa/pull/348 which needs to be merged first. It implements transformers derived from OpenStreetMap into the workflow. The transformer component is currently **one** transformer type chosen in the config that we apply for all voltage levels.

It is the foundation of choosing flexibly transformers depending on the voltage level changes between buses:
https://github.com/pypsa-meets-africa/pypsa-africa/issues/285

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by pz-max** [comment]
@ekatef FYI, this worked perfectly. Thanks for some previous work. 
Once your PR is merged, I can update this one & merge it as well.

File: scripts/build_osm_network.py
Comment (pz-max) [inline scripts/build_osm_network.py]: @davide-f there are a couple of probably unnecessary columns available.
I would remove for now:
- length
- circuits
- tag_type
- bounds
- bus_0_coors + bus_1_coors

But keeping all these columns, for now, is probably not problematic. Meaning we could also deal later with it... How do you feel about it?

Comment (davide-f) [reply scripts/build_osm_network.py]: Personally, I'd like to structure the csv files of transformers are they are supposed to be, similarly to pypsa-eur.
This is also to highlight possible issues and avoid possible problems down the line.

**Comment by davide-f** [comment]
@pz-max could you plot the (a) raw osm data, (b) cleaned osm data, (c) base_network and (d) simplify_network networks using the new transformer implementation?

This is to make sure that the workflow properly works without loosing pieces along the line and avoiding to have disconnected networks. That is impossible to be automatically checked with the current CI

**Comment by davide-f** [comment]
Anyway, before this PR, we shall merge https://github.com/pypsa-meets-africa/pypsa-africa/pull/348 that is almost ready anyway

File: scripts/build_osm_network.py
Comment (pz-max) [reply scripts/build_osm_network.py]: Mh, I think carrying the geojson might be good for plotting and the voltage_bus0 and voltage_bus1 are good for choosing voltage-specific transformers. All other values can be dropped. But yeah the voltage can be also easy derived from matching with the bus list

Comment (davide-f) [reply scripts/build_osm_network.py]: No on the voltages and location I totally agree: it is way easier to have them there for plotting; the rest can be dropped

**Comment by pz-max** [comment]
> @pz-max could you plot the (a) raw osm data, (b) cleaned osm data, (c) base_network and (d) simplify_network networks using the new transformer implementation?
> 
> This is to make sure that the workflow properly works without loosing pieces along the line and avoiding to have disconnected networks. That is impossible to be automatically checked with the current CI

## Raw osm network
There is a substation from which power lines go out.
raw_osm_network_kano

## Clean osm network
Every line gets a substation bus at the ending.
clean_osm_kano

## Build osm network 
The network is meshed since lines that are connected in a substation can transfer power across each other.
build_network_kano

## Base network
Red line represents a transformer, which means the voltage level between some lines are different. Works also quite well at other locations.
base_network_kano

## Simplify network
Removes non-circuits:
simplify_network

## Simplify network over Nigeria
simplify_network-fullNG

## osm build network across Nigeria
osm_built_network-fullNG

**Comment by davide-f** [comment]
Perfect! As the PR by katia is merged, I will promptly merge this as well :)

**Comment by pz-max** [comment]
@davide-f seems to work perfectly with the staged PR :)

**Review by davide-f — APPROVED** [review]

**Comment by davide-f** [comment]
> @davide-f seems to work perfectly with the staged PR :)

Just checking the new merge conflict, there has been an update by katia, sees the last commit. Feel free to comment or adapt. As we have green light, this PR will be merged

**Comment by pz-max** [comment]
> > @davide-f seems to work perfectly with the staged PR :)
> 
> Just checking the new merge conflict, there has been an update by katia, sees the last commit. Feel free to comment or adapt. As we have green light, this PR will be merged

Changes are ok. Green light on the merge_conflict resolution. Let's wait for the CI....

**Pull request merged**
